### PR TITLE
support clusters that depend on NLB target groups

### PR DESCRIPTION
**Pull request by robfletcher**
Even though we can't manage NLBs we can detect their target groups and not reject clusters that depend on them.

### CODE REVIEW

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/AmazonLoadBalancer.kt [hunk 1]
```diff
@@ -24,6 +25,4 @@ interface AmazonLoadBalancer {
   val vpcId: String
   val subnets: Set<String>
   val scheme: String?
-  val idleTimeout: Int
-  val securityGroups: Set<String>
```
Comment (robfletcher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/AmazonLoadBalancer.kt:28]: Had to move these fields off the interface as NLBs don't have them. Didn't require any other changes.

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/NetworkLoadBalancerModel.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  override val vpcId: String,
+  override val subnets: Set<String>,
+  override val scheme: String?,
+  override val availabilityZones: Set<String>,
+  val targetGroups: List<TargetGroup>
+) : AmazonLoadBalancer {
+  data class TargetGroup(
+    val targetGroupName: String
+  )
+}
```
Comment (robfletcher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/NetworkLoadBalancerModel.kt:18]: Basic NLB model. We just want to check that the thing exists and has the target group we want

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredLoadBalancerVeto.kt [hunk 3]
```diff
@@ -73,7 +74,13 @@ class RequiredLoadBalancerVeto(
     }
     spec.targetGroupsByRegion.forEach { (name, account, regions) ->
       val missingRegions = regions.filter { region ->
-        none { lb -> lb is ApplicationLoadBalancerModel && lb.account == account && lb.region == region && lb.targetGroups.any { it.targetGroupName == name } }
+        none { lb ->
+          when (lb) {
+            is ApplicationLoadBalancerModel -> lb.account == account && lb.region == region && lb.targetGroups.any { it.targetGroupName == name }
+            is NetworkLoadBalancerModel -> lb.account == account && lb.region == region && lb.targetGroups.any { it.targetGroupName == name }
+            else -> false
+          }
```
Comment (robfletcher) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredLoadBalancerVeto.kt:82]: Once again I wish Kotlin had intersection types

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/NetworkLoadBalancerModel.kt (hunk 2 above)
Comment (gal-yardeni) [reply keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/NetworkLoadBalancerModel.kt:18]: It's probably better to add here a comment that we don't support it yet? So we / anyone else looking at the code will be confused 🤷

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredLoadBalancerVeto.kt (hunk 3 above)
Comment (lorin) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredLoadBalancerVeto.kt:82]: if you were really bothered by the duplication here, you could derive an interface from `AmazonLoadBalancer` that has target groups on it (e.g., `TargetGroupedAmazonLoadBalancer`).

**Review by lorin — APPROVED** [review]
## My understanding

### Context

For reasons internal to Netflix, keel isn't able to automatically manage NLBs, so keel has never supported them.

### Problem

An internal user exported resources associated with an app. One of those resources was an EC2 cluster fronted by an NLB. The EC2 cluster got stuck in `MISSING_DEPENDENCY` state.

### Proposed Solution

Make keel aware of NLB resources by defining a `NetworkLoadBalancerModel` class. This class is deserialized from the `CloudDriverService.loadBalancersForApplication` RPC call for objects that have `name = "network"`

With this solution, even though keel can't actually manage NLB resources, keel` won't throw an error if an NLB is associated with a managed cluster.

## Feedback

The change looks good overall, but I don't understand the original failure mode well enough to judge whether this will fix the problem.  How was keel behaving when an EC2 cluster was fronted by NLB, and we did not support NLB model objects?

**Pull request merged**